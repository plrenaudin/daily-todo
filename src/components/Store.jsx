import React from "react";
import {
  LOAD,
  ADD,
  addTodo,
  REMOVE,
  removeTodo,
  listTodo,
  TOGGLE,
  toggleTodo,
  listWeekTodo
} from "../modules/model";
import { isoFormatDate, currentWeekISO } from "../modules/date";

const ConfigStore = React.createContext();
const TodoStore = React.createContext();
const WeekStore = React.createContext();

const todoReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case ADD:
      return addTodo(state, action.data);
    case REMOVE:
      return removeTodo(state, action.data);
    default:
      throw new Error();
  }
};

const weekReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case TOGGLE:
      return toggleTodo(state, action.todo, action.day);
    default:
      throw new Error();
  }
};

const useDayTodos = () => {
  const [week, dispatch] = React.useReducer(weekReducer, []);
  return { week, dispatch };
};

const useTodos = () => {
  const [todos, dispatch] = React.useReducer(todoReducer, []);
  return { todos, dispatch };
};

const Provider = ({ children }) => {
  //Config
  const [currenDate, setCurrentDate] = React.useState(isoFormatDate());
  const configStore = { currenDate, setCurrentDate };

  //Week
  const weekStore = useDayTodos();

  //Todos
  const todoStore = useTodos();

  React.useEffect(() => {
    const week = currentWeekISO();
    Promise.all([listTodo(), listWeekTodo(week)]).then(([todos, weekTodos]) => {
      todoStore.dispatch({ type: LOAD, data: todos });
      weekStore.dispatch({ type: LOAD, data: weekTodos });
    });
  }, []);

  return (
    <ConfigStore.Provider value={configStore}>
      <TodoStore.Provider value={todoStore}>
        <WeekStore.Provider value={weekStore}>{children}</WeekStore.Provider>
      </TodoStore.Provider>
    </ConfigStore.Provider>
  );
};

export { ConfigStore, WeekStore, TodoStore, Provider };
