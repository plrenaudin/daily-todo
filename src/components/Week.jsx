import React from "react";
import { WeekHeader } from "./WeekHeader";
import styled from "styled-components";
import { ConfigStore, TodoStore, WeekStore } from "./Store";
import { currentWeekISO } from "../modules/date";
import { TOGGLE, REMOVE } from "../modules/model";
import { TodoInput } from "./TodoInput";
import { generateColor } from "../modules/color";

const Table = styled.table`
  border-spacing: 0.3rem;
  border-collapse: separate;
`;

const Row = styled.tr`
  text-align: center;
`;

const Cell = styled.td`
  padding: 1rem 0;
  background-color: ${props => props.color || "transparent"};
`;
const TodoCell = styled(Cell)`
  text-align: left;
  padding: 0 0.3rem;
  background-color: ${props => props.color || "transparent"};
`;
const onToggle = (dailyTodoStore, day, todo) => () => {
  dailyTodoStore.dispatch({ type: TOGGLE, todo, day });
};

const Week = React.memo(() => {
  const config = React.useContext(ConfigStore);
  const todoStore = React.useContext(TodoStore);
  const dailyTodoStore = React.useContext(WeekStore);
  const remove = data =>
    window.confirm(`Delete ${data}?`) &&
    todoStore.dispatch({ type: REMOVE, data });

  return (
    <Table>
      <WeekHeader />
      <tbody>
        {todoStore.todos.map(todo => {
          const baseColor = generateColor(todo, 93);
          const doneColor = generateColor(todo, 70);

          return (
            <Row key={todo}>
              {currentWeekISO(config.currentDate).map(day => {
                const done = (dailyTodoStore.week[day] || []).includes(todo);
                return (
                  <Cell key={day} color={done ? doneColor : baseColor}>
                    <input
                      type="checkbox"
                      checked={done}
                      onChange={onToggle(dailyTodoStore, day, todo)}
                    />
                  </Cell>
                );
              })}
              <TodoCell color={doneColor}>
                <label>{todo}</label>{" "}
                <button onClick={() => remove(todo)}>X</button>
              </TodoCell>
            </Row>
          );
        })}
        <Row>
          {currentWeekISO(config.currentDate).map(day => (
            <Cell key={day}>&nbsp;</Cell>
          ))}
          <TodoCell>
            <TodoInput />
          </TodoCell>
        </Row>
      </tbody>
    </Table>
  );
});

export { Week };
