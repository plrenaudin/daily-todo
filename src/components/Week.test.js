import React from "react";
import renderer from "react-test-renderer";
import { Week } from "./Week";
import { ConfigStore, TodoStore, WeekStore } from "./Store";

it("renders the Week correctly", () => {
  const component = (
    <ConfigStore.Provider value={{ currentDate: "2019-02-24" }}>
      <TodoStore.Provider value={{ todos: ["plip", "huhu"] }}>
        <WeekStore.Provider value={{ week: { "2019-02-24": ["huhu"] } }}>
          <Week />
        </WeekStore.Provider>
      </TodoStore.Provider>
    </ConfigStore.Provider>
  );
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});
