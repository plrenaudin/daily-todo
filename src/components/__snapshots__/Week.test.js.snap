// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Week correctly 1`] = `
.c2 {
  padding: 1rem;
}

.c3 {
  font-size: 1.3rem;
  font-weight: bold;
  display: block;
}

.c4 {
  color: grey;
}

.c1 {
  padding: 0.3rem;
}

.c0 {
  border: 1px solid grey;
}

.c5 {
  background-color: lightgray;
}

.c6 {
  padding: 0.3rem;
}

<table
  className="c0"
>
  <thead>
    <tr
      className="c1"
    >
      <th
        className="c2"
      >
        <em
          className="c3"
        >
          Fr
        </em>
        <small
          className="c4"
        >
          22nd
        </small>
      </th>
      <th
        className="c2"
      >
        <em
          className="c3"
        >
          Sa
        </em>
        <small
          className="c4"
        >
          23rd
        </small>
      </th>
      <th
        className="c2"
      >
        <em
          className="c3"
        >
          Su
        </em>
        <small
          className="c4"
        >
          24th
        </small>
      </th>
      <th
        className="c2"
      >
        <em
          className="c3"
        >
          Mo
        </em>
        <small
          className="c4"
        >
          25th
        </small>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
    </tr>
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
      <td
        className="c6"
      >
        <input
          checked={true}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
      <td
        className="c6"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
